modules/ssh/init.go

```diff
@@ -0,0 +1,51 @@
+package ssh
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+func Init() error {
+	if setting.SSH.Disabled {
+		return nil
+	}
+
+	if setting.SSH.StartBuiltinServer {
+		Listen(setting.SSH.ListenHost, setting.SSH.ListenPort, setting.SSH.ServerCiphers, setting.SSH.ServerKeyExchanges, setting.SSH.ServerMACs)
+		log.Info("SSH server started on %s. Cipher list (%v), key exchange algorithms (%v), MACs (%v)",
+			net.JoinHostPort(setting.SSH.ListenHost, strconv.Itoa(setting.SSH.ListenPort)),
+			setting.SSH.ServerCiphers, setting.SSH.ServerKeyExchanges, setting.SSH.ServerMACs,
+		)
+		return nil
+	}
+
+	builtinUnused()
+
+	// FIXME: why 0o644 for a directory .....
+	if err := os.MkdirAll(setting.SSH.KeyTestPath, 0o644); err != nil {
+		return fmt.Errorf("failed to create directory %q for ssh key test: %w", setting.SSH.KeyTestPath, err)
+	}
+
+	if len(setting.SSH.TrustedUserCAKeys) > 0 && setting.SSH.AuthorizedPrincipalsEnabled {
+		caKeysFileName := setting.SSH.TrustedUserCAKeysFile
+		caKeysFileDir := filepath.Dir(caKeysFileName)
+
+		err := os.MkdirAll(caKeysFileDir, 0o700) // it should be the SSH.RootPath by default (`~/.ssh` in most cases)
```

Comment:
It's not necessarily the `.ssh/` directory. The `TrustedUserCAKeysFile` is actually referred to directly by the sshd configuration. We use `.ssh/` by default on docker but it's actually anywhere that the user's sshd.conf points to.

See #12224

--- 

This will need extra sshd configuration like this

```
AuthorizedPrincipalsFile .ssh/authorized_principals
TrustedUserCAKeys /etc/ssh/trusted-user-ca-keys.pem
```

---

That being said,  0o700 is ok as a default and if anyone needs it to be something different they should just create the damned directory first.

Reply:
Yup, I mean `by default`.

Reply:
```suggestion
		err := os.MkdirAll(caKeysFileDir, 0o700) // it might be the SSH.RootPath by default (`~/.ssh` in most cases)
```

I just left the comment for future readers to know what the directory usually should be.

Maybe my comment is accurate. How about `it might be`? Or do you have better suggestions?

Reply:
```suggestion
		err := os.MkdirAll(caKeysFileDir, 0o700) // SSH.RootPath by default (That is `~/.ssh` in most cases)
```